runner/websocket.go

```diff
@@ -0,0 +1,103 @@
+package runner
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"io"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pganalyze/collector/output/pganalyze_collector"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+	"google.golang.org/protobuf/proto"
+)
+
+func SetupWebsocketForAllServers(ctx context.Context, servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx := range servers {
+		go func(server *state.Server) {
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			for {
+				if server.WebSocket == nil {
+					connect(ctx, server, globalCollectionOpts, logger)
+				}
+				time.Sleep(60 * time.Second) // Delay between reconnect attempts
+			}
+		}(servers[idx])
+	}
+}
+
+func connect(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	connCtx, cancelConn := context.WithCancel(ctx)
+	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/websocket")
+	if url.Scheme == "http" {
+		url.Scheme = "ws"
+	} else {
+		url.Scheme = "wss"
+	}
+	headers := make(map[string][]string)
+	headers["Pganalyze-Api-Key"] = []string{server.Config.APIKey}
+	headers["Pganalyze-System-Id"] = []string{server.Config.SystemID}
+	headers["Pganalyze-System-Type"] = []string{server.Config.SystemType}
+	headers["Pganalyze-System-Scope"] = []string{server.Config.SystemScope}
+	headers["Pganalyze-System-Id-Fallback"] = []string{server.Config.SystemIDFallback}
+	headers["Pganalyze-System-Type-Fallback"] = []string{server.Config.SystemTypeFallback}
+	headers["Pganalyze-System-Scope-Fallback"] = []string{server.Config.SystemScopeFallback}
+	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
+	headers["Sec-WebSocket-Protocol"] = []string{"websocket"}
+	conn, response, err := websocket.DefaultDialer.DialContext(connCtx, url.String(), headers)
+	if err != nil {
+		cancelConn()
+		time.Sleep(10 * time.Second)
+		logger.PrintError("Error starting websocket: %s %v", err, response)
+		return
+	}
+	server.WebSocket = conn
+	go func() {
+		for {
+			select {
+			case <-connCtx.Done():
+				err := server.WebSocket.Close()
+				if err != nil {
+					logger.PrintError("Error closing websocket: %s", err)
+					return
+				}
+				server.WebSocket = nil
+				return
+			case snapshot := <-server.SnapshotStream:
```

Comment:
I wonder if it would be nice to either have a separate stream or pass type in addition to the snapshot bytes data, so that we will know which snapshot we're processing here. (e.g. to have a log of which snapshot went well and which didn't, like we log currently).
Also `logger.PrintError` is used for the below log but I think it's not an error level log

Reply:
In 8e0234d0bc7d536f5750d094ac90d0eccf6d10d4 I updated the code to print similar log lines to what you see with the existing API. The `PrintError` I was using for testing; I've changed that to `PrintVerbose`.